### ISSUE TITLE
Implementing a Socket Server

### ISSUE THREAD

**Issue by pwilsoncentre** [krakjoe/parallel#59]
Apologies in advance if this request shows a fundamental misunderstanding of the intended use for this extension.

I'm trying to envision how one would go about creating a socket server that can listen for connections using stream_socket_accept in a parallels\Event loop, notify a worker thread in a pool, continue listening for connections while a worker thread generates a response, and reply to a given client when the event loop is notified of a response by a worker thread.

In my initial experimentation, I'm running into a few things. If I listen for connections without blocking, the loop becomes very busy and fully utilizes a CPU core. If I do block while waiting for connections, I become unable to react to response events from worker threads. If I do the socket handling in a new thread and send events to a parent event loop, I can't pass socket or stream resources via a channel to be able to handle the socket/stream in a worker.

I feel like I'm misunderstanding something fundamentally. Ultimately, my goal here is to implement persistent worker threads which handle PSR-7 HTTP requests. My inspiration is https://roadrunner.dev/, which (if you aren't familiar with it) implements the actual socket server in Go and communicates with persistent worker PHP threads via a socket or STDIN/STDOUT.

Any examples or guidance would be greatly appreciated.

**Comment by dktapps** [comment]
If you want to listen without blocking, you'll need to add some kind of delay in your loop so it doesn't just spin constantly. Typically the way I would do this is to poll the socket every X milliseconds, and between polls do whatever other things you need to do.

Given that parallel doesn't currently have any way to do a select() which can be interrupted from another thread (for example with an IPC socket), I think polling is your only option.

**Comment by krakjoe** [comment]
I think if there is a fundamental misunderstanding, it's that the Events loop in parallel is for non-blocking I/O: It definitely is not for that.

There are already very good event loops for non-blocking I/O, and I suggest you use one of those, if threads can be useful to your application, it will not be in performing nbio.

Noteworthy, you can (in *nix) pass a stream over a channel - it will be cast to int, which you can use to fopen the stream again in another thread .... although I can't really recommend that you use that, because it will never work in windows.

Having said all of that, there is obviously a use case for offloading work in your described application, but you should keep the I/O and CPU loads separate by design.

**Comment by derekbsnider** [comment]
Has anyone been successful with this idea? I've been playing around with the same kind of idea, but I cannot have a different parallel\Runtime for each connection, because you cannot share the controlling socket between Runtimes (it's a resource), and you cannot bind to the same port in multiple threads.

**Comment by pwilsoncentre** [comment]
I have begun looking at other options. There's an interesting project out there that uses a socket server written in Go that spawns PHP worker threads which loop, waiting for the Go supervisor to send them requests via STDIN or a socket, which the PHP worker threads then execute. It's called RoadRunner:
https://roadrunner.dev/

It has its own limitations, the foremost in my opinion being that it can't handle streaming responses yet, although it's in their roadmap. As it stands now, the full response body gets sent to the Go supervisor before being passed along to the user. Once a response is generated, the PHP worker thread blocks and waits for another request from the supervisor. Another issue is that while it's handling a request the PHP worker thread can't handle other requests, so IO-bound applications will probably have some issues, although configuring the supervisor to spawn more workers might help.

The holy grail, in my opinion, would be to have PHP worker threads with their own event loops so that a single worker can handle many requests at a time. I think, in theory, this could be done natively in PHP using a master process that runs a socket server accepting HTTP requests and then sends the requests via STDIN to child processes, each of which runs an event loop that will process the requests and send them back to the master process via STDOUT or a socket. I don't have the time to attempt such an exercise, and I'm not even sure what sort of performance we would see from such a system. I'm afraid that the master process' socket server would be the bottleneck at that point.

**Comment by Frago9876543210** [comment]
> Noteworthy, you can (in *nix) pass a stream over a channel - it will be cast to int, which you can use to fopen the stream again in another thread .... although I can't really recommend that you use that, because it will never work in windows.

What about socket_create? You can't pass it into runtime like stream resource

**Comment by tobiastromm** [comment]
Has anyone been successful with this idea? [2]

**Comment by derekbsnider** [comment]
I gave up on this idea, and instead started working on my own C-language
variant, that is parsed and JIT compiled on the fly.

It will support some select features from other languages (C++, PHP, Perl,
Python, JavaScript), contained within namespaces.

So like you can do "using namespace std" to access standard C++ classes
without prefixing with "std::", you could do "using namespace php" to do
the same for the included PHP features, maybe like php::array, but I'd also
make it so that php::array was a class that had all the array functions as
class member functions.

Once I get the basic (bare bones) C/C++ functionality working stable, I'll
put it up on GitHub.

**Comment by TheTechsTech** [comment]
Have you looked into **coroutines** (https://github.com/symplely/coroutine), there is an article written by **Nikita Popov** [Cooperative multitasking using coroutines (in PHP!)](https://nikic.github.io/2012/12/22/Cooperative-multitasking-using-coroutines-in-PHP.html) he did some years ago, an project was started then abandoned. I built upon, more of an rewrite. and separated out the socket server part into another repo https://github.com/symplely/socket. It still an work in progress. 

The thing i see that other languages have expanded upon is the use of `yield` the generator. The whole async stuff in **Go, Python, C#, C, Java, and Javascript** is embraced and built around the same concept. PHP has introduced `yield` back in 5.5, but users/developers on the whole seem to have not grasped the idea. It's usage was envisioned from [RFC](https://wiki.php.net/rfc/generator-delegation) creators. My project is trying to fix that, for general userland usage.

**Comment by Fabbok1x** [comment]
> Noteworthy, you can (in *nix) pass a stream over a channel - it will be cast to int, which you can use to fopen the stream again in another thread .... although I can't really recommend that you use that, because it will never work in windows.

 fopening an integer doesnt work. fopen(4, ".."). How would that be done?
